### ISSUE TITLE
SQL driver doesn't support timestamp (time.Time)

### ISSUE THREAD

**Issue by Darkclainer** [genjidb/genji#258]
I have seen an issue about supporting `time.Time` in sql driver: https://github.com/genjidb/genji/issues/154.

But it seems that `time.Time` still doesn't work. Parser doesn't recognize keyword `TIMESTAMP` (not surprisingly, no changes was made in driver in corresponding MR for that issue).

```go
package main

import (
	"database/sql"
	"fmt"
	"time"

	_ "github.com/genjidb/genji/sql/driver"
)

func main() {
	db, err := sql.Open("genji", ":memory:")
	if err != nil {
		panic(err)
	}
	defer db.Close()

	_, err = db.Exec(`CREATE TABLE foo (created_at TIMESTAMP NOT NULL)`)
	if err != nil {
		panic(err)
	}
       //...
```

Prints: 

```sh
panic: found TIMESTAMP, expected ) at line 1, char 30
```

Scanning `time.Time` stored as `TEXT` doesn't work either:

```go
package main

import (
	"database/sql"
	"fmt"
	"time"

	_ "github.com/genjidb/genji/sql/driver"
)

func main() {
	db, err := sql.Open("genji", ":memory:")
	if err != nil {
		panic(err)
	}
	defer db.Close()

	_, err = db.Exec(`CREATE TABLE foo (created_at TEXT NOT NULL)`)
	if err != nil {
		panic(err)
	}

	_, err = db.Exec(`INSERT INTO foo (created_at) VALUES (?)`, time.Now().UTC())
	if err != nil {
		panic(err)
	}

	rows, err := db.Query(`SELECT created_at FROM foo`)
	if err != nil {
		panic(err)
	}
	defer rows.Close()
	for rows.Next() {
		var createdAt time.Time
		if err := rows.Scan(&createdAt); err != nil {
			panic(err)
		}
		fmt.Printf("foo found: (%v)", createdAt)
	}
	if err := rows.Err(); err != nil {
		panic(err)
	}
}
```

Prints: 

```sh
panic: sql: Scan error on column index 0, name "created_at": unsupported Scan, storing driver.Value type string into type *time.Time
```

Is this a bug or I misusing sql driver?

**Comment by tdemin** [comment]
Can confirm, doesn't work for me either with both Genji 0.7.2 and 0.8.0.

**Comment by tdemin** [comment]
Running the latest Genji CLI to convert data to timestamp results in this:

```
genji> update tasks set ctime = cast(ctime as timestamp);
found timestamp, expected type at line 1, char 40
```

**Comment by asdine** [comment]
The timestamp support in Genji is currently very light, hopefully we will tackle this in the next release.

The `TIMESTAMP` type is not supported, only the conversion between `time.Time` and `TEXT` and vice versa.

However, the second example should be working. It does when using the low-level API, so this looks like a bug.

**Comment by tzzed** [comment]
@asdine `TIMESTAMP` should be supported as type or as alias? What is expected?

**Comment by asdine** [comment]
This issue should focus only on fixing the behaviour described here:

```go
package main

import (
	"database/sql"
	"fmt"
	"time"

	_ "github.com/genjidb/genji/sql/driver"
)

func main() {
	db, err := sql.Open("genji", ":memory:")
	if err != nil {
		panic(err)
	}
	defer db.Close()

	_, err = db.Exec(`CREATE TABLE foo (created_at TEXT NOT NULL)`)
	if err != nil {
		panic(err)
	}

	_, err = db.Exec(`INSERT INTO foo (created_at) VALUES (?)`, time.Now().UTC())
	if err != nil {
		panic(err)
	}

	rows, err := db.Query(`SELECT created_at FROM foo`)
	if err != nil {
		panic(err)
	}
	defer rows.Close()
	for rows.Next() {
		var createdAt time.Time
		if err := rows.Scan(&createdAt); err != nil {
			panic(err)
		}
		fmt.Printf("foo found: (%v)", createdAt)
	}
	if err := rows.Err(); err != nil {
		panic(err)
	}
}
```

```
panic: sql: Scan error on column index 0, name "created_at": unsupported Scan, storing driver.Value type string into type *time.Time
```

The goal is not to add support for the `TIMESTAMP` type.

**Comment by tzzed** [comment]
`rows.Scan(dest ...interface)` doesn't support `time.Time`.  There are many topics about the `time.Time` support.
https://github.com/go-sql-driver/mysql#timetime-support
https://stackoverflow.com/questions/29341590/how-to-parse-time-from-database

There are two ways to handle your need. 
See this notes. 
1. https://github.com/go-sql-driver/mysql/wiki/Examples#rawbytes

2. Use Genji sql's driver method `Scanner`.

https://github.com/genjidb/genji/blob/0ca41de8c5604b829f11bca60ccc00dc366ced8e/sql/driver/driver.go#L435-L439

```go
 // ... 
        for rows.Next() {
		var createdAt time.Time
		if err := rows.Scan(Scanner(&createdAt)); err != nil {
			panic(err)
		}
		fmt.Printf("foo found: (%v)", createdAt)
	}
        if err := rows.Err(); err != nil {
		panic(err)
	}
```

**Comment by asdine** [comment]
There was a bug preventing the use of a `time.Time` with `driver.Scanner`. This has been fixed by #384.
Because of `database/sql` limitations there is currently no way of scanning a `TEXT` field into a `time.Time` value without using `driver.Scanner`. Here is the correct way to do it:

https://play.golang.org/p/6k8xYN4Ghdj

```go
package main

import (
	"database/sql"
	"fmt"
	"time"

	"github.com/genjidb/genji/sql/driver"
)

func main() {
	db, err := sql.Open("genji", ":memory:")
	if err != nil {
		panic(err)
	}
	defer db.Close()

	_, err = db.Exec(`CREATE TABLE foo (created_at TEXT NOT NULL)`)
	if err != nil {
		panic(err)
	}

	_, err = db.Exec(`INSERT INTO foo (created_at) VALUES (?)`, time.Now().UTC())
	if err != nil {
		panic(err)
	}

	rows, err := db.Query(`SELECT created_at FROM foo`)
	if err != nil {
		panic(err)
	}
	defer rows.Close()
	for rows.Next() {
		var createdAt time.Time
		if err := rows.Scan(driver.Scanner(&createdAt)); err != nil {
			panic(err)
		}
		fmt.Printf("foo found: (%v)", createdAt)
	}
	if err := rows.Err(); err != nil {
		panic(err)
	}
}
```

Once the support for the `TIMESTAMP` will be implemented, `TIMESTAMP` fields will automatically translate to `time.Time` values without the use of `driver.Scanner`
